`extends` in `tsconfig.json` should prefer `tsconfig.json` over `package.json`

### Describe the bug

```jsonc
// lib/tsconfig.json
{
  "extends": "..", // it means `../tsconfig.json` actually, but `../package.json#main` is resolved unexpectedly
  "compilerOptions": {
    "rootDir": ".",
    "outDir": "../lib"
  }
}
```

```jsonc
// tsconfig.json
{
  "extends": "@1stg/tsconfig/node16"
}
```

```jsonc
// package.json
{
  "main": "lib/index.cjs"
}
```

```log
 FAIL  test/basic.spec.ts [ test/basic.spec.ts ]
Error: parsing /Users/JounQin/Workspaces/GitHub/test/lib/index.cjs failed: SyntaxError: Unexpected token ''', "'use stric"... is not valid JSON
 ❯ node_modules/vite/dist/node/chunks/dep-R0I0XnyH.js:13998:10

Caused by: SyntaxError: Unexpected token ''', "'use stric"... is not valid JSON
 ❯ node_modules/vite/dist/node/chunks/dep-R0I0XnyH.js:13990:24
```

### Reproduction

https://github.com/JounQin/test/tree/vitest_repo

### Steps to reproduce

```sh
yarn && yarn test
```

### System Info

```shell
System:
    OS: macOS 14.2
    CPU: (10) arm64 Apple M1 Max
    Memory: 15.73 GB / 64.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 20.10.0 - ~/Library/Caches/fnm_multishells/29289_1702804971977/bin/node
    Yarn: 4.0.2 - ~/Library/Caches/fnm_multishells/29289_1702804971977/bin/yarn
    npm: 10.2.3 - ~/Library/Caches/fnm_multishells/29289_1702804971977/bin/npm
    bun: 1.0.18 - /opt/homebrew/bin/bun
  Browsers:
    Brave Browser: 120.1.61.100
    Chrome: 120.0.6099.109
    Safari: 17.2
```


### Used Package Manager

yarn

### Logs

_No response_

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitejs/vite/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitejs.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitejs/vite/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a Vite issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to [vuejs/core](https://github.com/vuejs/core) instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite/discussions) or join our [Discord Chat Server](https://chat.vitejs.dev/).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/packages/vite/package.json
+++ b/packages/vite/package.json
@@ -135,7 +135,7 @@
     "source-map-support": "^0.5.21",
     "strip-ansi": "^7.1.0",
     "strip-literal": "^2.0.0",
-    "tsconfck": "^3.0.0",
+    "tsconfck": "^3.0.1",
     "tslib": "^2.6.2",
     "types": "link:./types",
     "ufo": "^1.3.2",

--- a/packages/vite/src/node/plugins/esbuild.ts
+++ b/packages/vite/src/node/plugins/esbuild.ts
@@ -444,14 +444,7 @@ async function loadTsconfigJsonForFile(
   filename: string,
 ): Promise<TSConfigJSON> {
   try {
-    if (tsconfckCache) {
-      // shortcut, the cache stores resolved TSConfckParseResult
-      // so getting it from the cache directly we bypass async fn call wrapping it in a promise again
-      if (tsconfckCache.hasParseResult(filename)) {
-        const result = await tsconfckCache.getParseResult(filename)
-        return result.tsconfig
-      }
-    } else {
+    if (!tsconfckCache) {
       tsconfckCache = new TSConfckCache<TSConfckParseResult>()
     }
     const result = await parse(filename, {

